build/lib/src/asset/id.dart

```diff
@@ -22,13 +22,57 @@ class AssetId implements Comparable<AssetId> {
   /// Gets the file extension of the asset, if it has one, including the ".".
   String get extension => pathos.extension(path);
 
-  /// Creates a new AssetId at [path] within [package].
+  /// Creates a new [AssetId] at [path] within [package].
   ///
   /// The [path] will be normalized: any backslashes will be replaced with
   /// forward slashes (regardless of host OS) and "." and ".." will be removed
   /// where possible.
   AssetId(this.package, String path) : path = _normalizePath(path);
 
+  /// Creates a new [AssetId] from an [uri].
+  ///
+  /// This gracefully handles `package:` URIs, which is useful when creating an
+  /// [AssetId] from an `import` or `export` directive pointing to a package:
+  /// ```dart
+  /// AssetId assetOfDirective(UriReferencedElement element) {
+  ///   return new AssetId.resolve(element.uri);
+  /// }
+  /// ```
+  ///
+  /// Optionally, specifying the origin asset ([from]) will also allow resolving
+  /// relative to another asset - otherwise an [ArgumentError] will be thrown.
+  /// ```dart
+  /// AssetId assetOfDirective(AssetId origin, UriReferencedElement element) {
+  ///   return new AssetId.resolve(element.uri, from: origin);
+  /// }
+  /// ```
+  factory AssetId.resolve(String uri, {AssetId from}) {
+    final parsedUri = Uri.parse(uri);
+    // If a dart:, package:, or file: URI was used, assume an absolute path.
+    if (parsedUri.hasScheme) {
+      if (parsedUri.scheme == 'package') {
+        return new AssetId(
+          parsedUri.pathSegments.first,
+          pathos.joinAll(parsedUri.pathSegments.skip(1)),
```

Comment:
Nit: we don't use trailing comments in this package for constructors. It's really unfortunate that this "feature" introduced effectively multiple styles 😢.